### PR TITLE
WIP climber subsystem extenders

### PR DESCRIPTION
**Pull request by BeastLaptop07**
Need to verify encoder values and speed the climber should move at.
Need to figure out CANSparkMax can command position, currently using set soft position limit and set(double speed)
PR open by: Nithila Balaji, Emily Enright, Maddy Paradis

tl edit: 2/5 added todo's from the code.
- [ ] Split into separate commajds
- [ ] State diagram
- [ ] Fully integrate (i.e. extend then pull back)

### CODE REVIEW

File: src/main/java/frc/robot/commands/Climbernetwork.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package frc.robot.commands;
+
+import com.ctre.phoenix.motorcontrol.can.MotControllerJNI;
+
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.NetworkTableEntry;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+public class Climbernetwork extends CommandBase {
```
Comment (tomlazar) [inline src/main/java/frc/robot/commands/Climbernetwork.java]:
```suggestion
public class TestClimber extends CommandBase {
```

This should eventually be able to run the climber in the simpliest way possible, if it becomes to complex to run just over network tables, then a series of commands should be used, or even switching to "production" quality commands.

File: src/main/java/frc/robot/commands/Climbernetwork.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package frc.robot.commands;
```
Comment (tomlazar) [inline src/main/java/frc/robot/commands/Climbernetwork.java]:
```suggestion
package frc.robot.commands.test;
```

should be in the test folder.

File: src/main/java/frc/robot/commands/Climbernetwork.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+public class Climbernetwork extends CommandBase {
+    private NetworkTable table;
+    private NetworkTableEntry left_extender_speed, right_extender_speed, left_extender_position, right_extender_position;
+    private Double last = 0.0;
+
+    // constructor to create network table
+    Climbernetwork(){
```
Comment (tomlazar) [inline src/main/java/frc/robot/commands/Climbernetwork.java]: match class name

File: src/main/java/frc/robot/commands/Climbernetwork.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        // remember last value to compare to new value and change if inputed value is different
+        last = 0.0;
+        left_extender_speed.setDouble(last);
+        right_extender_speed.setDouble(last);
+        left_extender_position.setDouble(last);
+        right_extender_position.setDouble(last);
+
+    }
+    @Override
+    public void execute(){
```
Comment (aaron97neu) [inline src/main/java/frc/robot/commands/Climbernetwork.java]: Needs to be finished. Look at this file for a rough example: https://github.com/2202Programming/FRC2022/blob/fullshooter/src/main/java/frc/robot/commands/test/TestMotorCommand.java

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    public boolean readyToTraverse() {
+        return false;
+    }
+    // @param inches from extender current position
+    public void setExtension(double inches) {
+        float count =  (float) inches; // convert command inches to encoder counts
+        SoftLimitDirection direction = null ;
+        double speed = 0 ;
```
Comment (tomlazar) [inline src/main/java/frc/robot/subsystems/Climber.java]: Rather then leaving this in a open loop control, we prefer to use PID loops and closed control systems running on the actual device for a more fine grained control.

For a example of a position PID loop see this [demo project](https://github.com/REVrobotics/SPARK-MAX-Examples/tree/master/Java/Position%20Closed%20Loop%20Control) by REV. Specifically this [line](https://github.com/REVrobotics/SPARK-MAX-Examples/blob/c18eb6f15f9d9b8d371da83e7302470f0d50c170/Java/Position%20Closed%20Loop%20Control/src/main/java/frc/robot/Robot.java#L112).

**Review by tomlazar — CHANGES_REQUESTED** [review]
@dlaufen we should probably remove the Mecanum parts at some point.

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    public String rung() {
+        return "blank";
+    }
+
+    public boolean readyToTraverse() {
+        return false;
+    }
+    // @param inches from extender current position
```
Comment (aaron97neu) [inline src/main/java/frc/robot/subsystems/Climber.java]: We need to reconsider if it is smarter to base our 0 off the current position or fully retracted. I am starting to think 0 being set at fully retracted is the best bet. All values should be positive in this case, need to enforce that.

**Review by aaron97neu — CHANGES_REQUESTED** [review]
Did review with Tom, agree with his requested changes. Added a few of my own

File: src/main/java/frc/robot/subsystems/Climber.java (hunk 5 above)
Comment (aaron97neu) [reply src/main/java/frc/robot/subsystems/Climber.java]: @dlaufen Could you link some example code that shows how to change from native units (rotations/pulses?) to sane (inches in this case). Would be a nice complement to add while making this change

**Comment by aaron97neu** [comment]
State diagram brainstormed 2/7

**Comment by tomlazar** [comment]
@aaron97neu 

```java

class ArmRotation {
    private Counter m_counter;
    private PWMMotorController m_motor;
    private double speed;
    private double tolerence;

    private int absPositon = 0;
    private int desPositon = 0;
    private Boolean kForward = true;

    public ArmRotation(Counter m_counter, PWMMotorController m_motor, double speed, double tolerence) {
        this.m_counter = m_counter;
        this.m_motor = m_motor;
        this.speed = speed;
        this.tolerence = tolerence;
    }


    public void periodic() {
        // add the adjustment
        double factor = 1;
        if (!kForward) factor = -1;
        absPositon += factor * m_counter.get();
        m_counter.reset();

        if(Math.abs(desPositon - absPositon ) > tolerence) {
            m_motor.set(factor * speed);
        } else {
            m_motor.set(0);
        }
    }

    public void set(int desired) {
        kForward = desired > absPositon;
        desPositon = desired;
    }
}
```

**Comment by aaron97neu** [comment]
Current status: 
- Climber Subsystem seems to be fully implemented. Still needs a second pass to clean stuff up. Specifically Extension routine should be abstracted to a class like ClimberRotation.
- Climber test file is created but needs to be implemented. Planning on finishing implementing and testing Wednesday with @nathanaelren if he is available

**Comment by aaron97neu** [comment]
Suggested plan of attack for the new limit switches:

- Communicate with mechanical team and identify limit switches, their type, location, how they are wired and document. I think they are all magnetic but it should not matter if that changes

- Assuming they are wired to the canspark max's, use this example to implement them: https://github.com/REVrobotics/SPARK-MAX-Examples/blob/master/Java/Limit%20Switch/src/main/java/frc/robot/Robot.java. The comments explain how things work nicely. Ensure you are using `SparkMaxLimitSwitch` objects/types. Additional docs: https://codedocs.revrobotics.com/java/com/revrobotics/cansparkmax

- If they are wired back to the rio example code from last year can be found here: https://github.com/2202Programming/FRC-2019/blob/master/FRC-Robot-2019/src/main/java/frc/robot/subsystems/ClimberSubsystem.java

Once that is finished, Climber subsystem needs to have a good cleanup and everything needs to be tested

**Comment by dlaufen** [comment]
Makes sense.  I expect limit switches will be directly wired to sparkmax.

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 7]
```diff
@@ -295,29 +246,158 @@ public void periodic() {
...
-        this.sdb_desired = table.getEntry("desired");
+    private SparkMaxLimitSwitch VerticalLimitSwitch;
+// nts
+    private NetworkTable network_table;
+    private NetworkTableEntry nte_curr_pos_deg;
+    private NetworkTableEntry nte_curr_pos_count;
+    private NetworkTableEntry nte_des_pos_deg;
+    private NetworkTableEntry nte_des_pos_count;
+    private NetworkTableEntry nte_curr_current_amp;
+    private NetworkTableEntry nte_curr_current_limit_amp;
```
Comment (tomlazar) [inline src/main/java/frc/robot/subsystems/Climber.java]: you_can_really_tell_aaron_only_programs_python

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 8]
```diff
@@ -295,29 +246,158 @@ public void periodic() {
...
+        nte_curr_pos_deg.setDouble(0);
+        nte_curr_pos_count = network_table.getEntry("Current Arm Rotation (counts)");
+        nte_curr_pos_count.setDouble(0);
+        nte_des_pos_deg = network_table.getEntry("Desired Arm Rotation (degrees)");
+        nte_des_pos_deg.setDouble(0);
+        nte_des_pos_count = network_table.getEntry("Desired Arm Rotation (counts)");
+        nte_des_pos_count.setDouble(0);
+        nte_curr_current_amp = network_table.getEntry("Current Amperage (amps)");
+        nte_curr_current_amp.setDouble(0);
+        nte_curr_current_limit_amp = network_table.getEntry("Amperage Limit (amps)");
```
Comment (tomlazar) [inline src/main/java/frc/robot/subsystems/Climber.java]: maybe group inputs/outputs under

`$root/input/whatever`
`$root/output/whatever`

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 9]
```diff
@@ -295,29 +246,158 @@ public void periodic() {
...
+        nte_curr_current_limit_amp.setDouble(0);
+
+        CONVERSION_FACTOR = motor_rot.getEncoder().getPositionConversionFactor();
```
Comment (tomlazar) [inline src/main/java/frc/robot/subsystems/Climber.java]: can this change when its runnin?

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 10]
```diff
@@ -295,29 +246,158 @@ public void periodic() {
...
     }
     public void periodic() {
-
+        double motor_curr_deg = motor_rot.getEncoder().getPosition();
+        nte_curr_pos_deg.setDouble(motor_curr_deg);
+        nte_curr_pos_count.setDouble(motor_curr_deg / CONVERSION_FACTOR);
+        nte_curr_current_amp.setDouble(motor_rot.getOutputCurrent());
+// current limit n stuff
+        motor_rot.setSmartCurrentLimit((int) nte_curr_current_limit_amp.getDouble(5));
```
Comment (tomlazar) [inline src/main/java/frc/robot/subsystems/Climber.java]: can trafic, test for changes

File: src/main/java/frc/robot/commands/MidClimbRetract.java [hunk 11]
```diff
@@ -22,11 +23,11 @@ public void execute() {
...
     @Override
-    // TODO: Velocity better than position because it allows for some error?
     public boolean isFinished() {
-        return ((Math.abs(climber.getLeftEncoder().getPosition()) <= Constants.ClimbSettings.TOLERANCE_LENGTH)
-               && (Math.abs(climber.getRightEncoder().getPosition()) <= Constants.ClimbSettings.TOLERANCE_LENGTH)
-        // TODO: Add checks for rotation also
+        return ((Math.abs(climber.getLeftExtEncoder().getPosition()) <= Constants.ClimbSettings.TOLERANCE_LENGTH)
+               && (Math.abs(climber.getRightExtEncoder().getPosition()) <= Constants.ClimbSettings.TOLERANCE_LENGTH)
+               && (Math.abs(climber.getRightRotEncoder().getPosition()) <= Constants.ClimbSettings.TOLERANCE_ROTATION)
+               && (Math.abs(climber.getLeftRotEncoder().getPosition()) <= Constants.ClimbSettings.TOLERANCE_LENGTH)
```
Comment (tomlazar) [inline src/main/java/frc/robot/commands/MidClimbRetract.java]: use the new subsystem method?

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 12]
```diff
@@ -295,29 +246,158 @@ public void periodic() {
...
+        return right_motor_rot.getEncoder();
+    }
+
+    public void setAmperageLimit(int limit) {
+        right_motor_ext.setSmartCurrentLimit((int)extension_amps.getDouble(limit));
+        left_motor_ext.setSmartCurrentLimit((int)extension_amps.getDouble(limit));
+    }
+
+    public boolean checkIsFinished(double ext_pos, double rot_pos) {
+        return(Math.abs(this.getLeftExtEncoder().getPosition() - ext_pos) <= Constants.ClimbSettings.TOLERANCE_LENGTH
```
Comment (tomlazar) [inline src/main/java/frc/robot/subsystems/Climber.java]: take tolerence in method, and move checks to extension/rotation subclass

**Pull request merged**